inc/c_util/rc_ptr.h

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#ifndef RC_PTR_H
+#define RC_PTR_H
+
+#ifdef __cplusplus
+#include <cstddef>
+#include <cstdarg>
+#else
+#include <stdarg.h>
+#include <stddef.h>
+#endif
+
+#include "c_pal/thandle.h"
+
+#include "umock_c/umock_c_prod.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+typedef void (*RC_PTR_FREE_FUNC)(void* context);
+
+typedef struct RC_PTR_TAG
+{
+    const void* ptr;
+    RC_PTR_FREE_FUNC free_func;
+} RC_PTR;
+
+THANDLE_TYPE_DECLARE(RC_PTR);
+
+#define PRI_RC_PTR "p"
+
+#define RC_PTR_VALUE(rc) ((rc)->ptr)
+
+#define RC_PTR_AS_VOIDPTR(rc) (((rc) == NULL) ? NULL : (rc)->ptr)
```

Comment:
typedefs, defines can stay out of the umock_c_prod area where ideally only MOCKABLE_FUNCTIUONS should exist.